File: algorithms.cpp
// ============================================================
// BFS & DFS Algorithm Implementation in C++
// This is the reference C++ source. The same logic is
// mirrored in algorithms.js for direct browser execution.
// To compile: g++ -std=c++17 algorithms.cpp -o algorithms
// ============================================================

#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_map>
#include <string>

using namespace std;

// Graph represented as adjacency list
unordered_map<int, vector<int>> graph;

// BFS - Breadth First Search
vector<int> bfs(int start, int totalNodes) {
    vector<bool> visited(totalNodes, false);
    vector<int> order;
    queue<int> q;

    visited[start] = true;
    q.push(start);

    while (!q.empty()) {
        int node = q.front();
        q.pop();
        order.push_back(node);

        for (int neighbor : graph[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }
    return order;
}

// DFS - Depth First Search (iterative)
vector<int> dfs(int start, int totalNodes) {
    vector<bool> visited(totalNodes, false);
    vector<int> order;
    stack<int> s;

    s.push(start);

    while (!s.empty()) {
        int node = s.top();
        s.pop();

        if (!visited[node]) {
            visited[node] = true;
            order.push_back(node);

            // Push neighbors in reverse for left-to-right traversal
            for (int i = graph[node].size() - 1; i >= 0; i--) {
                if (!visited[graph[node][i]]) {
                    s.push(graph[node][i]);
                }
            }
        }
    }
    return order;
}

int main() {
    // Example: Algorithm Learning Roadmap graph
    graph[0] = {1, 2};       // Start -> Arrays, Strings
    graph[1] = {3, 4};       // Arrays -> Sorting, Searching
    graph[2] = {4, 5};       // Strings -> Searching, Hashing
    graph[3] = {6};           // Sorting -> Trees
    graph[4] = {6, 7};       // Searching -> Trees, Graphs
    graph[5] = {7};           // Hashing -> Graphs
    graph[6] = {8};           // Trees -> Dynamic Programming
    graph[7] = {8};           // Graphs -> Dynamic Programming
    graph[8] = {};             // Dynamic Programming (end)

    cout << "BFS Order: ";
    for (int n : bfs(0, 9)) cout << n << " ";
    cout << endl;

    cout << "DFS Order: ";
    for (int n : dfs(0, 9)) cout << n << " ";
    cout << endl;

    return 0;
}

File: algorithms.js
// ============================================================
// algorithms.js — BFS & DFS (mirrors algorithms.cpp logic)
// Uses same data structures: Queue (array+shift) = std::queue
//                            Stack (array+pop)   = std::stack
// ============================================================

/**
 * BFS — Breadth First Search
 * Returns full traversal order AND parent map for path tracing
 */
function bfs(start, graph, totalNodes) {
  const visited = new Array(totalNodes).fill(false);
  const parent  = new Array(totalNodes).fill(-1);
  const order   = [];
  const queue   = [];   // std::queue equivalent

  visited[start] = true;
  queue.push(start);

  while (queue.length > 0) {
    const u = queue.shift();   // front + pop
    order.push(u);

    for (const v of (graph[u] || [])) {
      if (!visited[v]) {
        visited[v]  = true;
        parent[v]   = u;
        queue.push(v);
      }
    }
  }
  return { order, parent };
}

/**
 * DFS — Depth First Search (iterative)
 * Returns full traversal order AND parent map for path tracing
 */
function dfs(start, graph, totalNodes) {
  const visited = new Array(totalNodes).fill(false);
  const parent  = new Array(totalNodes).fill(-1);
  const order   = [];
  const stack   = [];   // std::stack equivalent

  stack.push(start);

  while (stack.length > 0) {
    const u = stack.pop();   // top + pop

    if (!visited[u]) {
      visited[u] = true;
      order.push(u);

      const neighbors = graph[u] || [];
      for (let i = neighbors.length - 1; i >= 0; i--) {
        const v = neighbors[i];
        if (!visited[v]) {
          parent[v] = u;
          stack.push(v);
        }
      }
    }
  }
  return { order, parent };
}

/**
 * Reconstruct path from start to end using parent array
 */
function getPath(end, parent) {
  const path = [];
  let cur = end;
  while (cur !== -1) {
    path.unshift(cur);
    cur = parent[cur];
  }
  return path;
}
